options/target_types.go

```diff
@@ -0,0 +1,152 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package options
+
+import (
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/descriptorpb"
+)
+
+type optionsType[T any] interface {
+	*T
+	proto.Message
+	GetFeatures() *descriptorpb.FeatureSet
+	GetUninterpretedOption() []*descriptorpb.UninterpretedOption
+}
+
+type elementType[OptsStruct any, Opts optionsType[OptsStruct]] interface {
+	proto.Message
+	GetOptions() Opts
+}
+
+type targetType[Elem elementType[OptsStruct, Opts], OptsStruct any, Opts optionsType[OptsStruct]] struct {
+	t                       descriptorpb.FieldOptions_OptionTargetType
+	setUninterpretedOptions func(opts Opts, uninterpreted []*descriptorpb.UninterpretedOption)
+	setOptions              func(elem Elem, opts Opts)
+}
+
+var (
+	targetTypeFile = newTargetType[*descriptorpb.FileDescriptorProto](
+		descriptorpb.FieldOptions_TARGET_TYPE_FILE, setUninterpretedFileOptions, setFileOptions,
+	)
+	targetTypeMessage = newTargetType[*descriptorpb.DescriptorProto](
+		descriptorpb.FieldOptions_TARGET_TYPE_MESSAGE, setUninterpretedMessageOptions, setMessageOptions,
+	)
+	targetTypeField = newTargetType[*descriptorpb.FieldDescriptorProto](
+		descriptorpb.FieldOptions_TARGET_TYPE_FIELD, setUninterpretedFieldOptions, setFieldOptions,
+	)
+	targetTypeOneof = newTargetType[*descriptorpb.OneofDescriptorProto](
+		descriptorpb.FieldOptions_TARGET_TYPE_ONEOF, setUninterpretedOneofOptions, setOneofOptions,
+	)
+	targetTypeExtensionRange = newTargetType[*descriptorpb.DescriptorProto_ExtensionRange](
+		descriptorpb.FieldOptions_TARGET_TYPE_EXTENSION_RANGE, setUninterpretedExtensionRangeOptions, setExtensionRangeOptions,
+	)
+	targetTypeEnum = newTargetType[*descriptorpb.EnumDescriptorProto](
+		descriptorpb.FieldOptions_TARGET_TYPE_ENUM, setUninterpretedEnumOptions, setEnumOptions,
+	)
+	targetTypeEnumValue = newTargetType[*descriptorpb.EnumValueDescriptorProto](
+		descriptorpb.FieldOptions_TARGET_TYPE_ENUM_ENTRY, setUninterpretedEnumValueOptions, setEnumValueOptions,
+	)
+	targetTypeService = newTargetType[*descriptorpb.ServiceDescriptorProto](
+		descriptorpb.FieldOptions_TARGET_TYPE_SERVICE, setUninterpretedServiceOptions, setServiceOptions,
+	)
+	targetTypeMethod = newTargetType[*descriptorpb.MethodDescriptorProto](
+		descriptorpb.FieldOptions_TARGET_TYPE_METHOD, setUninterpretedMethodOptions, setMethodOptions,
+	)
+)
+
+func newTargetType[Elem elementType[OptsStruct, Opts], OptsStruct any, Opts optionsType[OptsStruct]](
```

Comment:
This factory function is here to reduce boiler-plate of the above var definitions: the Go compiler can infer type arguments for generic functions, but not for struct literal expressions for generic structs.